### PR TITLE
Feature/image carousel—it looks dope!

### PR DESCRIPTION
**Pull request by jkao1**
image
image

### CODE REVIEW

File: src/js/modules/RoutingApp.js [hunk 1]
```diff
@@ -181,12 +180,6 @@ class RoutingApp extends Component {
                 key={"latest"}
                 component={LatestPage}
               />
-              <Route
-                exact
-                path={"/search"}
-                key={"search"}
-                component={SearchPage}
-              />
               <Route
```
Comment (jkao1) [inline src/js/modules/RoutingApp.js]: I deleted all Search functionality to keep this branch pure. The search actions/components will be in @JasonLin43212's next PR.

File: dist/dev/index.html [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
-    <meta charset="utf-8">
-    <link href="/css/bootstrap.min.css" rel="stylesheet" type='text/css'>
-    <link href="/css/main-index.css" rel="stylesheet" type="text/css"/>
-    <link href='/css/main-fonts.css' rel='stylesheet' type='text/css' />
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <meta charset="utf-8">
+  <link href="/css/bootstrap.min.css" rel="stylesheet" type='text/css'>
+  <link href="/css/main-index.css" rel="stylesheet" type="text/css"/>
+  <link href='/css/main-fonts.css' rel='stylesheet' type='text/css' />
+  <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/slick-carousel/1.6.0/slick.min.css" />
```
Comment (NicholasLYang) [inline dist/dev/index.html]: Yeah...nope. Download the CSS file

Comment (NicholasLYang) [reply dist/dev/index.html]: We should actually package all of them tbh

File: src/js/constants.js [hunk 3]
```diff
@@ -13,4 +13,5 @@ export const MEDIA_CREATOR_SLUGS = {
 };
 
 export const EMAIL_REGEX = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i;
-export const SPEC_REEFER_PATTERN = /<spec-reefer id=(\d*)><\/spec-reefer>/;
\ No newline at end of file
+export const SPEC_REEFER_PATTERN = /<spec-reefer id=(\d*)><\/spec-reefer>/;
```
Comment (NicholasLYang) [inline src/js/constants.js]: Reefer?

File: src/js/modules/articles/components/ArticleBody.js [hunk 4]
```diff
@@ -1,14 +1,22 @@
...
-import { SPEC_REEFER_PATTERN } from "../../../constants";
+import {
+  SPEC_REEFER_PATTERN,
+  SPEC_IMG_CAROUSEL_PATTERN,
+} from "../../../constants";
 
 import ArticleFeaturedMedia from "./ArticleFeaturedMedia";
 import RightRail from "./RightRail";
 
+import Gallery from "../../media/components/Gallery";
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: Why the inconsistent imports?

File: src/js/modules/articles/components/ArticleBody.js [hunk 5]
```diff
@@ -119,35 +152,65 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
   //  featuredMedia = Object.values(articleMedia).find(image => image.isFeatured);
   const generateArticleReefer = content => {
     const match = SPEC_REEFER_PATTERN.exec(content);
-    const article = articles[parseInt(match[1])];
-    console.log(article.title)
-    if (article) {
-      const title = article.title.replace('“', '‘').replace('”', '’');
+    const reeferArticle = articles[parseInt(match[1])];
+    if (reeferArticle) {
+      const title = reeferArticle.title.replace("“", "‘").replace("”", "’");
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: Sorta confused by this. Maybe give a better name than just `title`?

File: src/js/modules/articles/components/ArticleBody.js [hunk 6]
```diff
@@ -119,35 +152,65 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
...
-          ,&rdquo; published in Volume {article.volume} Issue{" "}
-          {article.issue}.
+          ,&rdquo; published in Volume {reeferArticle.volume} Issue{" "}
+          {reeferArticle.issue}.
         </span>
       );
     }
   };
   const featuredMedia = media.find(medium => medium.isFeatured);
+  const displayImgCarousel = SPEC_IMG_CAROUSEL_PATTERN.test(content);
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: Ehhh. Not a great name for a boolean

File: src/js/modules/articles/components/ArticleBody.js [hunk 7]
```diff
@@ -119,35 +152,65 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
...
   return (
     <Row>
       <Col xs={12} sm={12} md={8} lg={8} className={classes.ArticleBody}>
-        {featuredMedia && (
-          <ArticleFeaturedMedia
-            featuredMedia={featuredMedia}
-            isCaption={true}
-          />
+        {displayImgCarousel ? (
+          Object.values(media).length > 0 && (
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: That condition should be included in the `displayImgCarousel`

File: src/js/modules/articles/components/ArticleFeaturedMedia.js [hunk 8]
```diff
@@ -48,7 +56,7 @@ class ArticleFeaturedMedia extends Component {
       imgWidth: 0,
     };
   }
-  componentDidMount() {
+  componentDidMount = () => {
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleFeaturedMedia.js]: Careful. Does this actually work? React lifecycle functions aren't like normal class functions

File: src/js/modules/articles/components/ArticleFeaturedMedia.js [hunk 9]
```diff
@@ -57,11 +65,17 @@ class ArticleFeaturedMedia extends Component {
         imgWidth: img.width,
       });
     };
-  }
-  render() {
-    const { classes, featuredMedia, isCaption, users } = this.props;
+  };
+  getComponent = key => {
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleFeaturedMedia.js]: I'm a little confused by this. Why is the enclosing component deciding which children to render? Why isn't a parent component deciding which children to give the enclosing component

File: src/js/modules/articles/components/ArticleFeaturedMedia.js [hunk 10]
```diff
@@ -57,11 +65,17 @@ class ArticleFeaturedMedia extends Component {
...
+    if (!this.props.children) {
+      return null;
+    }
+    return this.props.children.find(child => child.key === key);
+  };
+  render = () => {
+    const { classes, featuredMedia, users } = this.props;
     const creator = users[featuredMedia.userId];
-
+    const carouselButton = this.getComponent("carouselButton");
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleFeaturedMedia.js]: What

File: src/js/modules/articles/components/ArticleList.js [hunk 11]
```diff
@@ -51,15 +51,18 @@ const ArticleList = ({ classes, articles, title, label }) => {
   }
   return (
     <Grid className={classes.ArticleList}>
-      {title && (
-        title === "VOICES" ? (
+      {title &&
+        (title === "VOICES" ? (
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleList.js]: Ehh. That's interesting. Hard coded?

File: src/js/modules/core/components/Lightbox.js [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    left: 0,
+    padding: "100px 30px 30px 60px",
+    position: "fixed",
+    top: 0,
+    width: "100%!important",
+    zIndex: 10000,
+  },
+  contentContainer: {
+    display: "flex",
+    "-webkit-box-align": "center",
```
Comment (NicholasLYang) [inline src/js/modules/core/components/Lightbox.js]: JSS vendor prefixing

File: src/js/modules/core/components/Lightbox.js [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+      display: "none",
+    },
+  },
+};
+
+const Lightbox = ({
+  classes,
+  children,
+  title,
+  lightboxIsOpen,
```
Comment (NicholasLYang) [inline src/js/modules/core/components/Lightbox.js]: What. Why. isLightboxOpen

File: src/js/modules/core/components/Lightbox.js [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+        </div>
+      </div>
+    );
+  } else {
+    return <span className="lightbox-in-hiding" />;
+  }
+};
+
+const mapStateToProps = state => ({
+  lightboxIsOpen: state.core.lightboxIsOpen,
```
Comment (NicholasLYang) [inline src/js/modules/core/components/Lightbox.js]: Terrible name

File: src/js/modules/media/components/Gallery.js [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    return (
+      <div className={classes.Gallery}>
+        <Slider
+          {...settings}
+          ref={c => (this.slider = c)}
+          className={classes.Slider}
+        >
+          {media.map(image => {
+            return (
+              <div className={classes.slideContainer} key={image.id}>
```
Comment (NicholasLYang) [inline src/js/modules/media/components/Gallery.js]: Sorta feel like this should be its own component

File: src/js/modules/media/components/Gallery.js [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+                  <img src={image.attachmentUrl} />
+                  <figcaption className={classes.slideCaption}>
+                    <p>{image.title}</p>
+                  </figcaption>
+                </figure>
+              </div>
+            );
+          })}
+        </Slider>
+        <div className={classes.sliderFooter}>
```
Comment (NicholasLYang) [inline src/js/modules/media/components/Gallery.js]: Eh. Also maybe a component. Could be inlined, but a name would be nice here

File: src/js/modules/newspapers/components/NewspaperPage.js [hunk 17]
```diff
@@ -1,2 +1 @@
-
-// make a component, NewspaperPage, that takes volume & issue props and does what i drew on the whiteboard
\ No newline at end of file
+// make a component, NewspaperPage, that takes volume & issue props and does what i drew on the whiteboard
```
Comment (NicholasLYang) [inline src/js/modules/newspapers/components/NewspaperPage.js]: TODO

File: src/js/modules/articles/components/ArticleBody.js (hunk 4 above)
Comment (jkao1) [reply src/js/modules/articles/components/ArticleBody.js]: wdym

File: src/js/modules/articles/components/ArticleFeaturedMedia.js (hunk 8 above)
Comment (jkao1) [reply src/js/modules/articles/components/ArticleFeaturedMedia.js]: Yes, I tested it. Basically, if the featured media is more vertical than horizontal, it'll be floated to the left of the body text and not span the whole width. You can see that behavior in the article page.

File: src/js/modules/articles/components/ArticleList.js (hunk 11 above)
Comment (jkao1) [reply src/js/modules/articles/components/ArticleList.js]: Yea. There's only one subsection with an art banner right now.

File: src/js/modules/core/components/Lightbox.js (hunk 13 above)
Comment (jkao1) [reply src/js/modules/core/components/Lightbox.js]: Took it from python conventions. What makes more sense?
"if the lightbox is open,..." or
"if is lightbox open...".

File: src/js/modules/articles/components/ArticleBody.js (hunk 4 above)
Comment (NicholasLYang) [reply src/js/modules/articles/components/ArticleBody.js]: Not a named import

File: src/js/modules/media/components/Gallery.js (hunk 15 above)
Comment (jkao1) [reply src/js/modules/media/components/Gallery.js]: I shouldn't split up Gallery; the actions & callbacks will get too confusing. It'll be all over the place.

File: src/js/modules/articles/components/ArticleBody.js [hunk 18]
```diff
@@ -117,39 +127,47 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
   }
   */
   //  featuredMedia = Object.values(articleMedia).find(image => image.isFeatured);
-  const generateArticleReefer = content => {
-    const match = SPEC_REEFER_PATTERN.exec(content);
-    const article = articles[parseInt(match[1])];
-    console.log(article.title)
-    if (article) {
-      const title = article.title.replace('“', '‘').replace('”', '’');
+  const generateReferencedArticle = content => {
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: Why is this not a component

File: src/js/modules/articles/components/ArticleBody.js [hunk 19]
```diff
@@ -117,39 +127,47 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
...
+          ,&rdquo; published in Volume {referencedArticle.volume} Issue{" "}
+          {referencedArticle.issue}.
         </span>
       );
     }
   };
-  const featuredMedia = media.find(medium => medium.isFeatured);
+  const isCarouselButtonVisible =
+    SPEC_IMG_CAROUSEL_PATTERN.test(content) && Object.values(media).length > 0;
+  console.log(media);
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: Remove console logs

File: src/js/modules/articles/components/ArticleFeaturedMedia.js [hunk 20]
```diff
@@ -44,102 +64,41 @@ class ArticleFeaturedMedia extends Component {
...
-    };
+    img.src = this.props.image.attachmentUrl;
+    img.onload = () =>
+      this.setState({isImageTall: img.height > img.width * 1.2});
   }
   render() {
-    const { classes, featuredMedia, isCaption, users } = this.props;
-    const creator = users[featuredMedia.userId];
-
+    const { classes, image, isCarouselButtonVisible, carouselImageCount, openLightbox } = this.props;
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleFeaturedMedia.js]: 80 col rule

File: src/js/modules/core/components/Lightbox.js [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+const Lightbox = ({
+  classes,
+  children,
+  title,
+  isLightboxVisible,
+  closeLightbox,
+}) => {
+  if (isLightboxVisible) {
```
Comment (NicholasLYang) [inline src/js/modules/core/components/Lightbox.js]: Why is this the lightbox's job? Why not just not render lightbox entirely?

File: src/js/modules/media/components/Gallery.js [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+                  <img src={image.attachmentUrl} /> 
+                  <figcaption className={classes.slideCaption}> 
+                    <p>{image.title}</p> 
+                  </figcaption> 
+                </figure> 
+              </div> 
+            )
+            })}
+        </Slider>
+        <div className={classes.sliderFooter}>
```
Comment (NicholasLYang) [inline src/js/modules/media/components/Gallery.js]: Make this a component. I really don't like how long this file is

File: src/js/modules/articles/components/ArticleReference.js [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+  },
+  "@media (max-width: 767px)": {
+    ArticleReference: {
+      padding: "0 2%",
+    },
+  },
+};
+
+const ArticleReference = ({ classes, articles, sections, articleId }) => {
+  if (!(articleId in articles)) {
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleReference.js]: Don't have ArticleReference check this. Have its parent check it

File: src/js/modules/articles/components/ArticleBody.js [hunk 24]
```diff
@@ -117,39 +99,30 @@ const ArticleBody = ({ classes, articles, sections, content, media }) => {
...
+  const isCarouselButtonVisible =
+    SPEC_IMG_CAROUSEL_PATTERN.test(content) && Object.values(media).length > 0;
+  const referencedArticleId = SPEC_REFERENCE_PATTERN.test(content)
+    ? parseInt(SPEC_REFERENCE_PATTERN.exec(content)[1])
+    : -1;
   return (
     <Row>
       <Col xs={12} sm={12} md={8} lg={8} className={classes.ArticleBody}>
-        {featuredMedia && (
+        {isCarouselButtonVisible && (
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleBody.js]: I'm not following this boolean's logic. Why does the carousel button's visibility determine the lightbox's visibility? Shouldn't it be the other way around?

File: src/js/modules/articles/components/ArticleFeaturedMedia.js [hunk 25]
```diff
@@ -44,65 +64,48 @@ class ArticleFeaturedMedia extends Component {
...
-              {creator.firstName}
-              {creator.lastName !== "" && " " + creator.lastName}
-            </Link>
-            .
-          </figcaption>
-        )}
+        <div className={classes.imgContainer}>
+          <img className={classes.img} src={image.attachmentUrl} />
+          {isCarouselButtonVisible && (
+            <button className={classes.lightboxButton} onClick={openLightbox}>
```
Comment (NicholasLYang) [inline src/js/modules/articles/components/ArticleFeaturedMedia.js]: Carousel button or lightbox button

**Pull request merged**